app-backend/tile/src/main/scala/image/SingleBandMosaic.scala

```diff
@@ -0,0 +1,338 @@
+import geotrellis.spark._
+import geotrellis.spark.io._
+import geotrellis.raster.GridBounds
+import geotrellis.slick.Projected
+import geotrellis.vector.{Extent, Polygon}
+import geotrellis.proj4._
+import geotrellis.raster.histogram._
+import cats.data._
+import cats.implicits._
+import java.util.UUID
+
+import com.azavea.rf.common.utils.TileUtils
+
+import scala.concurrent._
+import geotrellis.spark.io.postgres.PostgresAttributeStore
+import com.azavea.rf.tile.AkkaSystem
+import com.typesafe.scalalogging.LazyLogging
+import geotrellis.raster.render.{ColorRamp, ColorMap}
+
+object SingleBandMosaic extends LazyLogging with KamonTrace {
+  lazy val memcachedClient = LayerCache.memcachedClient
+  implicit val database = Database.DEFAULT
+
+  val system = AkkaSystem.system
+  implicit val blockingDispatcher =
+    system.dispatchers.lookup("blocking-dispatcher")
+
+  val store = PostgresAttributeStore()
+  val rfCache = new CacheClient(memcachedClient)
+
+  def tileLayerMetadata(id: UUID, zoom: Int)(
+    implicit database: Database, sceneIds: Set[UUID]
+  ): OptionT[Future, (Int, TileLayerMetadata[SpatialKey])] = {
+    logger.debug(s"Requesting tile layer metadata (layer: $id, zoom: $zoom")
+    LayerCache.maxZoomForLayer(id).mapFilter {
+      case (pyramidMaxZoom) =>
+        val layerName = id.toString
+        for (maxZoom <- pyramidMaxZoom.get(layerName)) yield {
+          val z = if (zoom > maxZoom) maxZoom else zoom
+          blocking {
+            z -> store.readMetadata[TileLayerMetadata[SpatialKey]](LayerId(layerName, z))
+          }
+        }
+    }
+  }
+
+  /** MultiBandMosaic tiles from TMS pyramids given that they are in the same projection.
+    *   If a layer does not go up to requested zoom it will be up-sampled.
+    *   Layers missing color correction in the mosaic definition will be excluded.
+    */
+  def apply(
+    project: Project, zoom: Int, col: Int, row: Int
+  )(implicit database: Database): OptionT[Future, MultibandTile] =
+    traceName(s"SingleBandMosaic.apply(${project.id})") {
+      logger.debug(s"Creating single band mosaic (project: ${project.id}, zoom: $zoom, col: $col, row: $row)")
+
+      val polygonBbox: Projected[Polygon] = TileUtils.getTileBounds(zoom, col, row)
+      project.singleBandOptions match {
+        case Some(singleBandOptions: SingleBandOptions.Params) =>
+          val futureTiles: Future[Seq[(MultibandTile, Array[Histogram[Double]])]] = database.db.run {
+            ScenesToProjects.filter(_.projectId === project.id).map(_.sceneId).result
+          } flatMap {
+            sceneIds => {
+              Future.sequence(
+                sceneIds map { sceneId =>
+                  rfCache.cachingOptionT(s"tile-${sceneId}-${zoom}-${col}-${row}")(
+                    SingleBandMosaic.fetch(sceneId, zoom, col, row)(database, sceneIds.toSet)
+                  ).value.zip(
+                    rfCache.cachingOptionT(s"layer-histogram-${sceneId}-${zoom}")(
```

Comment:
This should be redundant - we should be able to just call `LayerCache.layerHistogram(sceneId, zoom)`.